service/crt/decoder.go

```diff
@@ -0,0 +1,38 @@
+package crt
+
+import (
+	"encoding/json"
+	microerror "github.com/giantswarm/microkit/error"
+
+	"github.com/giantswarm/certificatetpr"
+	"k8s.io/client-go/pkg/runtime"
+	"k8s.io/client-go/pkg/watch"
+)
+
+// certificateDecoder is used by the watcher to decode the certificatetpr objects.
+type certificateDecoder struct {
+	decoder *json.Decoder
+	close   func() error
+}
+
+// Decode deserializes a runtime object into a certificatetpr CustomObject.
+func (cd *certificateDecoder) Decode() (watch.EventType, runtime.Object, error) {
+	var e struct {
+		Type   watch.EventType
+		Object certificatetpr.CustomObject
+	}
+
+	if err := cd.decoder.Decode(&e); err != nil {
+		return watch.Error, nil, microerror.MaskAny(err)
+	}
+
+	return e.Type, &e.Object, nil
+}
+
+// Close closes the stream.
+func (cd *certificateDecoder) Close() {
+	err := cd.close()
+	if err != nil {
+		panic(err)
+	}
+}
```

Comment:
In PR description:

> the decoder did not close the watch stream.

I think the only change that is meaningful here is reusing json.Decoder. Stream instance was closed in previous implementation.
I'd like to make that clear, that's why I'm suggesting creating constructor function, like

```
func newcertificateDecoder(stream io.ReadCloser) {
    return certificateDecoder struct {
         decoder:  decoder: json.NewDecoder(stream),
         stream: stream,
     }
}
```